tests/test_roles.py

```diff
@@ -0,0 +1,757 @@
+        assert child.equals(copied_child)
+
+    def test_filter_user(self, update, role, parent_role):
+        update.message.chat = None
+        assert not role(update)
+
+        role.add_member(0)
+        assert role(update)
+
+        update.message.from_user.id = 1
+        assert not role(update)
+
+        parent_role.add_child_role(role)
+        parent_role.add_member(1)
+        assert role(update)
+
+    def test_filter_chat(self, update, role, parent_role):
+        update.message.from_user = None
+        assert not role(update)
+
+        role.add_member(0)
+        assert role(update)
+
+        update.message.chat.id = 1
+        assert not role(update)
+
+        parent_role.add_child_role(role)
+        parent_role.add_member(1)
+        assert role(update)
+
+    def test_filter_merged_roles(self, update, role):
+        role.add_member(0)
+        r = Role(0)
+        assert not (role & (~r))(update)
+
+        r = Role(1)
+        assert not (role & r)(update)
+        assert (role | r)(update)
+
+    def test_filter_allow_parent(self, update, role, parent_role):
+        role.add_member(0)
+        parent_role.add_member(1)
+        parent_role.add_child_role(role)
+
+        test_role = ~role
+        assert not test_role(update)
+        update.message.from_user.id = 1
+        update.message.chat.id = 1
+        assert test_role(update)
+
+    def test_filter_exclude_children(self, update, role, parent_role):
+        parent_role.add_child_role(role)
+        parent_role.add_member(0)
+        role.add_member(1)
+
+        test_role = ~parent_role
+        assert not test_role(update)
+        update.message.from_user.id = 1
+        update.message.chat.id = 1
+        assert not test_role(update)
+        update.message.from_user.id = 2
+        update.message.chat.id = 1
+        assert not test_role(update)
+
+    def test_filter_without_user_and_chat(self, update, role):
+        role.add_member(0)
+        update.message = None
+        assert not role(update)
+
+    def test_always_allow_admin(self, update, role):
+        role._admin.add_member(0)
+        try:
+            assert (~Role(0))(update)
+            assert (Role(0) & ~Role(0))(update)
+            assert (Role(1) & ~Role(0))(update)
+            assert (Role(1) & ~Role(2))(update)
+            assert (Role(0) | ~Role(0))(update)
+            assert (Role(1) | ~Role(0))(update)
+            assert (Role(1) | ~Role(2))(update)
+        finally:
+            role._admin.kick_member(0)
+
+    def test_pickle(self, role, parent_role):
+        role.add_member([0, 1, 3])
+        parent_role.add_member([4, 5, 6])
+        child_role = Role(name='child_role', chat_ids=[7, 8, 9])
+        role.add_child_role(child_role)
+        parent_role.add_child_role(child_role)
+
+        data = {
+            'role': role,
+            'parent': parent_role,
+            'child': child_role,
+        }
+        with open('pickle', 'wb') as file:
```

Comment:
https://stackoverflow.com/a/44608345/3827785
Would prefer not depending on file system (read: "please use bytesio")

Reply:
what's the downside?